### DISCUSSION TITLE
Как экспортировать анимации (.skl) с моделей в которых используется Armature Tools > Create Connected Bones > Create IK

### DISCUSSION THREAD

**Discussion by shadept-GitHub** [PavelBlend/blender-xray#463] [Q&A]
Я вот начал юзать новую фичу, которую завезли относительно недавно. Можно создавать контрол-риг с помощью этой фичи, но вот как экспортировать анимацию с настоящих костей в которой есть движения из контроль-рига я не знаю. Можете подсказать?
![image](https://user-images.githubusercontent.com/65095770/152867406-f1cc0857-8de2-45bc-827b-66a991bf43d8.png)

И ещё, как можно добавить анимацию в существующий файл .omf? У меня при экспорте в .omf блендер орёт что "armature object has no bone groups" хоть у самих костей всё есть и всё двигает модельку.

**Comment by shadept-GitHub** [comment]
К этому ещё добавилось "selected objecs have no actions" при экспорте .skl хоть я и запек анимацию в изначальный риг.

**Comment by PavelBlend** [comment]
@shadept-GitHub вот сделал пример:
[test_anim_export.zip](https://github.com/PavelBlend/blender-xray/files/8023155/test_anim_export.zip)


У экспортируемого объекта нужно указать Dependency Object:
![01](https://user-images.githubusercontent.com/7983249/152974502-fea0dfb6-20a4-42fd-b120-875db192cc4b.jpg)

Список анимаций во вкладке Motions (на скриншоте выше) должен быть заполнен у объекта original. У объекта connected можно не заполнять.

Объект original привязан к объекту connected с помощью оператора Link Bones (это делается автоматически после нажатия Create Connected Bones). Если нужно отвязать оригинальный скелет от connected скелета, то нужно нажать Unlink Bones:
![02](https://user-images.githubusercontent.com/7983249/152974977-5a467774-2c29-4dde-b163-cc08365529d1.jpg)

При экспорте нужно выделять объект original. Объект connected - это вспомогательный объект, который нужен только в блендере.

**Reply by shadept-GitHub** [reply]
Понятно. А как насчет этого?

> К этому ещё добавилось "selected objecs have no actions" при экспорте .skl хоть я и запек анимацию в изначальный риг.

Я вот через Replace и Add только смог экспортировать .omf при том, что существует файл .omf в папке экспорта с таким же именем, как и у импортированного. Но после импорта уже измененного .omf в пустую сцену с моделькой обычного сталкера в списке .omf нету новой анимации, которую я добавил через плюсик, в котором указал новый action. Может, эти манипуляции нужно делать в СДК? То есть, добавлять новые анимации через его в уже существующий .omf (я использовал stalker_scenario_animation.omf).

**Comment by PavelBlend** [comment]
> "selected objecs have no actions"

@shadept-GitHub  это означает, что экспортируемые объекты не имеют анимаций в списке motions. Нужно заполнить этот список:
![01](https://user-images.githubusercontent.com/7983249/153045722-a2523982-30dd-43d7-b0d8-f5096cfbd643.jpg)
Чтобы его быстро заполнить, нужно использовать маленькие кнопки с иконками справа от списка. Назначение кнопок:
![03](https://user-images.githubusercontent.com/7983249/153047380-28834614-418e-4b78-ac66-af02a8e567e6.jpg)


Экспорт skls и batch-экспорт skl анализируют этот список и экспортируют action'ы из него. Но для экспорта одной skl, нужно использовать этот оператор, а не из меню File > Export:
![02](https://user-images.githubusercontent.com/7983249/153045732-7dde9e55-e825-4c3e-aaef-ae0517fee29f.jpg)

**Comment by PavelBlend** [comment]
> У меня при экспорте в .omf блендер орёт что "armature object has no bone groups" хоть у самих костей всё есть и всё двигает модельку.

bone group - это не группы вершин. Они находятся здесь:
![4](https://user-images.githubusercontent.com/7983249/153048005-a11fac66-6793-4619-bc15-924d23aaeeb4.jpg)

**Reply by shadept-GitHub** [reply]
А какой их смысл? Их можно импортировать в блендер в сам же скелет модели?

**Reply by PavelBlend** [reply]
@shadept-GitHub это аналог Bone Parts в Actor Editor:
![01](https://user-images.githubusercontent.com/7983249/153060531-f9f69126-e8b7-4509-ad01-0ccdb64f2498.jpg)
Они нужны, чтобы разделить скелет на части. В персонажах обычно разделяются ноги, руки и голова, чтобы в игре проигрывать разные анимации для разных bone part.

**Reply by shadept-GitHub** [reply]
Выходит ошибка при экспорте .omf 
`Python: Traceback (most recent call last):
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\io_scene_xray\utils.py", line 487, in wrapper
    return method(self, context)
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\io_scene_xray\utils.py", line 508, in wrapper
    return method(self, context)
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\io_scene_xray\omf\ops.py", line 312, in execute
    exp.export_omf_file(export_context)
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\io_scene_xray\omf\exp.py", line 551, in export_omf_file
    chunked_writer = export_omf(context)
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\io_scene_xray\log.py", line 16, in wrap
    return func(*args, **kwargs)
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\io_scene_xray\omf\exp.py", line 453, in export_omf
    available_boneparts, available_params = get_motion_params(chunks[fmt.Chunks.S_SMPARAMS])
  File "C:\Users\User\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\io_scene_xray\omf\exp.py", line 126, in get_motion_params
    motion.writer.data.append(struct.pack('<I', num_marks))
TypeError: 'bytes' object cannot be interpreted as an integer

location: <unknown location>:-1`



Все, что я сделал это overwrite на существующий файл .omf, в который я добавил новую анимацию с Dependency, в котором я выбрал кости connected. Потом я их убрал и ошибка до сих пор выскакивает.

**Reply by shadept-GitHub** [reply]
Я вот пытался по этому туториалу добавить анимации в существующий файл .omf но в блендере. Я вот что не понял, а вообще нужно ли изменять существующий файл .omf или можно создать новый под другим именем. Если да то как его указать потом в игровых файлах state_lib.script или же state_mgr_animation_list.script чтобы анимацию скрипты брали из определенного файла .omf, который я создал https://youtu.be/hNbEewMyRSk

Кроме этого я не могу разобраться как правильно экспортировать Bone Parts из СДК. Когда я импортировал .object файл выведен из блендера, в котором я импортировал файл .ogf в окно Bone Parts то СДК просто вылетел. 

Может уже есть более новый туториал по созданию собственных анимаций для персонажей т.к. в ютубе мало видео на данную тему. СДК если что у меня ЗП. Репак от DaaGuda.

**_При экспорте .bones из СДК и импорте файла в блендер всплывает сообщение про то, что в файле нету информации по Bone Parts. Модельки в СДК и блендере одинаковые вроде бы._**

**Reply by shadept-GitHub** [reply]
Я ещё пробовал экспортировать .omf  файл из СДК со своим добавленным .skl экспортированным через спрятанную панель action-ов. 

В СДК лог весь зелёный без проблем, а вот как попытаться в блендер вставить то ошибку выдает что наверное там структура вся не та или что. В omfeditor-е тоже ошибку при загрузке дает. 

СДК все пробовал почти. И без патчей с moddb, и 64 бит и 32 бит DaaGuda. Ничего в итоге не выходит. Я просто взял в блендере и экспортировал модельку stalker_neutral_2 как .object и всё, вроде бы работает в СДК и в блендере тоже наверное.

**Reply by PavelBlend** [reply]
> Выходит ошибка при экспорте .omf

исправил. Пока релиз не делал. Фикс в последнем коммите.


> Я вот что не понял, а вообще нужно ли изменять существующий файл .omf или можно создать новый под другим именем.

зависит от ситуации. Иногда можно создать отдельный omf файл.




> Если да то как его указать потом в игровых файлах state_lib.script или же state_mgr_animation_list.script чтобы анимацию скрипты брали из определенного файла .omf, который я создал

я скриптами не занимаюсь, сам не знаю.



> Кроме этого я не могу разобраться как правильно экспортировать Bone Parts из СДК. Когда я импортировал .object файл выведен из блендера, в котором я импортировал файл .ogf в окно Bone Parts то СДК просто вылетел.

в блендере, после импорта ogf, нужно импортировать bone parts из omf с такими параметрами:
![01](https://user-images.githubusercontent.com/7983249/153166313-6c85800c-4e88-4e60-87d3-8eebc69de0ea.jpg)
После чего создадутся bone groups:
![02](https://user-images.githubusercontent.com/7983249/153166353-b49634a0-adc1-4ab7-a039-de63bad1618f.jpg)

**Reply by PavelBlend** [reply]
Вот здесь я подробнее описывал, как импортировать boneparts:
https://github.com/PavelBlend/blender-xray/issues/344#issuecomment-917685466

**Reply by shadept-GitHub** [reply]
Ясно, спасибо. Решил тем, что создал ОМФ файл в блендере с экспортированными Bone Parts с помощью AXRToolset. Не знал что можно с помощью блендера их из ОМФ файлов достать. Потом я просто связал два файла ОМФ (а именно тот, который я создал и stalker_scenario_animations и я также сразу пробовал импортировать stalker_scenario_animation в блендер и туда добавить анимацию свою и всё работало) и всё работает.
